nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -1681,13 +1992,123 @@ pub fn volume_is_read_only(
+///
+/// Note that UUIDs are not randomized by this step: Crucible will reject a
+/// `target_replace` call if the replacement VolumeConstructionRequest does not
+/// exactly match the original, except for a single Region difference.
+///
+/// Note that the generation number _is_ bumped in this step, otherwise
+/// `compare_vcr_for_update` will reject the update.
+fn replace_region_in_vcr(
+    vcr: &VolumeConstructionRequest,
+    old_region: SocketAddrV6,
+    new_region: SocketAddrV6,
+) -> anyhow::Result<VolumeConstructionRequest> {
+    match vcr {
+        VolumeConstructionRequest::Volume {
+            id,
+            block_size,
+            sub_volumes,
+            read_only_parent,
+        } => Ok(VolumeConstructionRequest::Volume {
+            id: *id,
+            block_size: *block_size,
+            sub_volumes: sub_volumes
+                .iter()
+                .map(|subvol| -> anyhow::Result<VolumeConstructionRequest> {
+                    replace_region_in_vcr(&subvol, old_region, new_region)
+                })
+                .collect::<anyhow::Result<Vec<VolumeConstructionRequest>>>()?,
+
+            // Only replacing R/W regions
+            read_only_parent: read_only_parent.clone(),
+        }),
+
+        VolumeConstructionRequest::Url { id, block_size, url } => {
+            Ok(VolumeConstructionRequest::Url {
+                id: *id,
+                block_size: *block_size,
+                url: url.clone(),
+            })
+        }
+
+        VolumeConstructionRequest::Region {
+            block_size,
+            blocks_per_extent,
+            extent_count,
+            opts,
+            gen,
+        } => {
+            let mut opts = opts.clone();
+
+            for target in &mut opts.target {
+                let parsed_target: SocketAddrV6 = target.parse()?;
+                if parsed_target == old_region {
+                    *target = new_region.to_string();
+                }
+            }
+
+            Ok(VolumeConstructionRequest::Region {
+                block_size: *block_size,
+                blocks_per_extent: *blocks_per_extent,
+                extent_count: *extent_count,
+                opts,
+                gen: *gen + 1,
+            })
+        }
+
+        VolumeConstructionRequest::File { id, block_size, path } => {
+            Ok(VolumeConstructionRequest::File {
+                id: *id,
+                block_size: *block_size,
+                path: path.clone(),
+            })
+        }
+    }
+}
+
+/// Find Regions in a Volume's subvolumes list whose target match the argument
+/// IP
+fn find_matching_rw_regions_in_volume(
+    vcr: &VolumeConstructionRequest,
+    ip: &std::net::Ipv6Addr,
+    targets: &mut Vec<SocketAddrV6>,
+) -> anyhow::Result<()> {
+    match vcr {
+        VolumeConstructionRequest::Volume { sub_volumes, .. } => {
+            for sub_volume in sub_volumes {
+                find_matching_rw_regions_in_volume(sub_volume, ip, targets)?;
```

Comment:
Unfortunately, the depth of any Volume Construction Request is arbitrary - customers can layer disks and snapshots as much as they want. I wasn't aware of a recursion limit in Rust! A quick search brings up [this in the Rust reference](https://doc.rust-lang.org/reference/attributes/limits.html#the-recursion_limit-attribute), which says it's 128 by default, but this seems like a compile time thing instead of a run time thing.

I can rework this (and other functions that recurse through VCRs) to be iterative if it's a problem, maybe as a follow up issue because this will end up affecting code not related to this PR.

Reply:
I'm fine doing it in a a follow up. But it seems relatively high priority.

Reply:
Opened #5815 :+1: